cpp/src/rolling/rolling.cu

```diff
@@ -626,5 +629,199 @@ std::unique_ptr<column> rolling_window(column_view const& input,
+      case cudf::TIMESTAMP_MILLISECONDS : return 24L*60*60*1000;
+      case cudf::TIMESTAMP_MICROSECONDS : return 24L*60*60*1000*1000;
+      default  : 
+        CUDF_EXPECTS(data_type.id() == cudf::TIMESTAMP_NANOSECONDS, 
+                    "Unexpected data-type for timestamp-based rolling window operation!");
+        return 24L*60*60*1000*1000*1000;
+    }
+  }
+
+  template <typename TimestampImpl_t>
+  std::unique_ptr<column> grouped_time_range_rolling_window_impl( column_view const& input,
+                                                      column_view const& timestamp_column,
+                                                      rmm::device_vector<cudf::size_type> const& group_offsets,
+                                                      rmm::device_vector<cudf::size_type> const& group_labels,
+                                                      size_type preceding_window_in_days, // TODO: Consider taking offset-type as type_id. Assumes days for now.
+                                                      size_type following_window_in_days,
+                                                      size_type min_periods,
+                                                      std::unique_ptr<aggregation> const& aggr,
+                                                      rmm::mr::device_memory_resource* mr) {
+
+    TimestampImpl_t mult_factor {static_cast<TimestampImpl_t>(multiplication_factor(timestamp_column.type()))};
+  
+    auto preceding_calculator = 
+      [
+        d_group_offsets = group_offsets.data().get(),
+        d_group_labels  = group_labels.data().get(),
+        d_timestamps    = timestamp_column.data<TimestampImpl_t>(),
+        preceding_window_in_days,
+        mult_factor
+      ] __device__ (size_type idx) {
+        auto group_label = d_group_labels[idx];
+        auto group_start = d_group_offsets[group_label];
+        auto lower_bound = d_timestamps[idx] - preceding_window_in_days*mult_factor;
+
+        return ((d_timestamps + idx) 
+                 - thrust::lower_bound(thrust::seq, d_timestamps + group_start, d_timestamps + idx, lower_bound)) 
+               + 1; // Add 1, for `preceding` to account for current row.
+      };
+  
+    auto following_calculator = 
+      [
+        d_group_offsets = group_offsets.data().get(),
+        d_group_labels  = group_labels.data().get(),
+        d_timestamps    = timestamp_column.data<TimestampImpl_t>(),
+        following_window_in_days,
+        mult_factor
+      ] __device__ (size_type idx) {
+        auto group_label = d_group_labels[idx];
+        auto group_end = d_group_offsets[group_label+1]; // Cannot fall off the end, since offsets is capped with `input.size()`.
+        auto upper_bound = d_timestamps[idx] + following_window_in_days*mult_factor;
+
+        return (thrust::upper_bound(thrust::seq, d_timestamps + idx, d_timestamps + group_end, upper_bound) 
+                 - (d_timestamps + idx))
+               - 1;
```

Comment:
I would suggest `thrust::prev` instead of `- 1` here, but it's only available in CUDA 10.1 and later so we can't yet.

Reply:
Noted. I'll have to remember this for later.